[pkg/stanza]: fix operator field config validation

braydonk opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
A few operators with `entry.Field` in their config were validating their configuration in their `Build` function via a comparison to `entry.NewNilField`. This was an inaccuracy; when a field is not provided during the unmarshal step, nothing ever gets run to actually construct the underlying struct field, meaning that the `entry.Field` was functionally set to `entry.Field{}`. This is not the same as `entry.NewNilField` which actually implements and supplies an interface value under the hood. This caused bad configs to pass validation and led to collector panic instead of nice failure messages.

This PR changed `entry.NewNilField` to actually produce the correct pattern for a "nil field", while leaving the original implementation intact in `entry.NewNoopField` in case it is useful elsewhere (though at the moment I only saw it used for nil comparisons, which as far as I can tell were either never useful or something got refactored without being caught). I also added functionality to
`operatortest.ConfigUnmarshalTest` to check for validation failures by expecting a particular error when calling the operator's `Build` method.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Addresses a panic discovered by #40398.

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Tests were added to each operator's config validation. I tested manually via a built collector by running the config from the linked issue and confirming that instead of `panic` I get an error message.
```shell
2025-06-15T01:53:10.582Z        info    service@v0.127.0/service.go:199 Setting up own telemetry...     {"resource": {}}
2025-06-15T01:53:10.582Z        info    builders/builders.go:26 Development component. May change in the future.        {"resource": {}, "otelcol.component.id": "debug", "otelcol.component.kind": "exporter", "otelcol.signal": "logs"}
Error: failed to build pipelines: failed to create "filelog/my-app-name" receiver for data type "logs": remove: field is empty
2025/06/15 01:53:10 collector server run finished with error: failed to build pipelines: failed to create "filelog/my-app-name" receiver for data type "logs": remove: field is empty
```

braydonk committed:
[pkg/stanza]: fix operator field config validation

A few operators with `entry.Field` in their config were validating their
configuration in their `Build` function via a comparison to
`entry.NewNilField`. This was an inaccuracy; when a field is not
provided during the unmarshal step, nothing ever gets run to actually
construct the underlying struct field, meaning that the `entry.Field`
was functionally set to `entry.Field{}`. This is not the same as
`entry.NewNilField` which actually implements and supplies an interface
value under the hood. This caused bad configs to pass validation and led
to collector panic instead of nice failure messages.

This PR changed `entry.NewNilField` to actually produce the correct
pattern for a "nil field", while leaving the original implementation
intact in `entry.NewNoopField` in case it is useful elsewhere (though at
the moment I only saw it used for nil comparisons, which as far as I can
tell were either never useful or something got refactored without being
caught). I also added functionality to
`operatortest.ConfigUnmarshalTest` to check for validation failures by
expecting a particular error when calling the operator's `Build` method.

braydonk committed:
Merge branch 'main' into operator_field_config_validation

braydonk committed:
fix lint errors

braydonk committed:
fix unit tests

Turns out the construction of the `field` config is different in certain
scenarios where all resource or all attributes are intended to be
removed, meaning the original comparison didn't cover the case properly.
This is now added.

Also fixed copy operator test configs to use a prefix that is actually
recognized so the field is parsed.

andrzej-stencel reviewed (commented):
Thank you very much for this, this fix is much needed. Please let me know what you think about the proposed implementation without the `Nil/NoopField` type.

braydonk commented on pkg/stanza/entry/noop_field.go:
```diff
@@ -0,0 +1,39 @@
+
+// NoopField is a struct that implements Field, but
+// does nothing for all its operations.
+type NoopField struct{}
```
I like the look of that. Let me mess with it, and if it works like I expect it would I will take both suggestions and delete the `NilField` and `NoopField` types entirely.

andrzej-stencel replied:
All right, great, thanks for this!

Deleting the `NilField` type is an API breaking change, how about we leave it in and deprecate it first? :sweat_smile: I suppose it doesn't hurt to leave it in in case someone is using it, right?

braydonk committed:
fix broken test and lint errors

braydonk committed:
remove test name prefix in TestProcessAndBuild since it's redundant

braydonk committed:
Remove NoopField, multi error validation, fix broken tests

This commit changes to a `Field.IsEmpty` to allow the deletion of the
`NoopField` and `NilField` types. It also adds multiple error validation
to allow for friendlier validation error messages for the `copy` and
`move` operators. Finally, it fixes some tests that were revealed to be
broken by relying on the previously faulty `copy` operator config
validation.

braydonk committed:
Merge branch 'main' into operator_field_config_validation

braydonk committed:
add changelog entry

braydonk committed:
Merge branch 'main' into operator_field_config_validation

andrzej-stencel reviewed (commented):
Awesome! Please let me know what you think about the breaking change concern

braydonk commented on pkg/stanza/entry/nil_field.go:
```diff
@@ -1,33 +0,0 @@
-// NilField is a struct that implements Field, but
-// does nothing for all its operations. It is useful
-// as a default no-op field to avoid nil checks.
-type NilField struct{}
```
Yeah good point; for some reason I'd been treating this type as if it was in an `internal` package this whole time.

Next time I have a sec to work on this I'll checkout the `nil_field.go` and `nil_field_test.go` from main and instead mark them as deprecated.

braydonk replied:
Sorry this fell through the cracks. Finally done in [d3196e7](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40728/commits/d3196e7c89f2e4a39c3131d93e606d084faf4b60)

braydonk committed:
restore NilField API and mark deprecated

braydonk committed:
Merge branch 'main' into operator_field_config_validation

andrzej-stencel committed:
Merge branch 'main' into operator_field_config_validation

andrzej-stencel reviewed (approved):
Thank you 🚀 👏 🙏
